crypto.c

```diff
@@ -385,40 +385,57 @@ static void dtls_ec_key_from_uint32(const uint32_t *key, size_t key_size,
   }
 }
 
-int dtls_ec_key_from_uint32_asn1(const uint32_t *key, size_t key_size,
-				 unsigned char *buf) {
-  int i;
-  unsigned char *buf_orig = buf;
-  int first = 1; 
-
-  for (i = (key_size / sizeof(uint32_t)) - 1; i >= 0 ; i--) {
-    if (key[i] == 0)
-      continue;
-    /* the first bit has to be set to zero, to indicate a poritive integer */
-    if (first && key[i] & 0x80000000) {
-      *buf = 0;
-      buf++;
-      dtls_int_to_uint32(buf, key[i]);
-      buf += 4;      
-    } else if (first && !(key[i] & 0xFF800000)) {
-      buf[0] = (key[i] >> 16) & 0xff;
-      buf[1] = (key[i] >> 8) & 0xff;
-      buf[2] = key[i] & 0xff;
-      buf += 3;
-    } else if (first && !(key[i] & 0xFFFF8000)) {
-      buf[0] = (key[i] >> 8) & 0xff;
-      buf[1] = key[i] & 0xff;
-      buf += 2;
-    } else if (first && !(key[i] & 0xFFFFFF80)) {
-      buf[0] = key[i] & 0xff;
-      buf += 1;
-    } else {
-      dtls_int_to_uint32(buf, key[i]);
-      buf += 4;
-    }
-    first = 0;
+/* Build the EC KEY as a ASN.1 positive integer */
+/*
+ * If the first non-zero byte of the 32 byte key has bit 7 set, then as this is
+ * negative, a leading zero byte needs to be prepended to make it a positive
+ * number for ASN.1.
+ *
```

Comment:
Alternative proposal:

The public EC key consist of two positive numbers. Converting them into ASN.1 INTEGER special care must be taken of the resulting sign. If the first non-zero byte of the 32 byte ec-key has bit 7 set (highest bit), the ASN.1 INTEGER would be interpreted mistakenly as negative number. In order to prevent from that mistake, a zero in the ASN.1 presentation is prepended, if that bit 7 is set.

Reply:
@boaks I like your alternative proposal.

Editorial nit: The commas in the last sentence should not be there.

Reply:
Updated, slightly re-worded.